src/hooks/useViewportColumns.ts

```diff
@@ -1,254 +1,40 @@
-    }
-
-    // get the first visible non-frozen column index
-    let colVisibleStartIdx = firstUnfrozenColumnIdx;
-    while (colVisibleStartIdx < lastColIdx) {
-      const { left, width } = columnMetrics.get(columns[colVisibleStartIdx])!;
-      // if the right side of the columnn is beyond the left side of the available viewport,
-      // then it is the first column that's at least partially visible
-      if (left + width > viewportLeft) {
-        break;
-      }
-      colVisibleStartIdx++;
-    }
-
-    // get the last visible non-frozen column index
-    let colVisibleEndIdx = colVisibleStartIdx;
-    while (colVisibleEndIdx < lastColIdx) {
-      const { left, width } = columnMetrics.get(columns[colVisibleEndIdx])!;
-      // if the right side of the column is beyond or equal to the right side of the available viewport,
-      // then it the last column that's at least partially visible, as the previous column's right side is not beyond the viewport.
-      if (left + width >= viewportRight) {
-        break;
-      }
-      colVisibleEndIdx++;
-    }
-
-    const colOverscanStartIdx = Math.max(firstUnfrozenColumnIdx, colVisibleStartIdx - 1);
-    const colOverscanEndIdx = Math.min(lastColIdx, colVisibleEndIdx + 1);
-
-    return [colOverscanStartIdx, colOverscanEndIdx];
-  }, [columnMetrics, columns, lastFrozenColumnIndex, scrollLeft, totalFrozenColumnWidth, viewportWidth, enableVirtualization]);
-
-  const viewportColumns = useMemo((): readonly CalculatedColumn<R, SR>[] => {
     const viewportColumns: CalculatedColumn<R, SR>[] = [];
     for (let colIdx = 0; colIdx <= colOverscanEndIdx; colIdx++) {
       const column = columns[colIdx];
 
-      if (colIdx < colOverscanStartIdx && !column.frozen) continue;
+      if (colIdx < startIdx && !column.frozen) continue;
       viewportColumns.push(column);
     }
 
     return viewportColumns;
-  }, [colOverscanEndIdx, colOverscanStartIdx, columns]);
-
-  return { columns, viewportColumns, layoutCssVars, columnMetrics, totalColumnWidth, lastFrozenColumnIndex, totalFrozenColumnWidth, groupBy };
-}
-
-function getSpecifiedWidth<R, SR>(
-  { key, width }: Column<R, SR>,
-  columnWidths: ReadonlyMap<string, number>,
-  viewportWidth: number
-): number | undefined {
-  if (columnWidths.has(key)) {
-    // Use the resized width if available
-    return columnWidths.get(key);
-  }
-  if (typeof width === 'number') {
-    return width;
-  }
-  if (typeof width === 'string' && /^\d+%$/.test(width)) {
-    return Math.floor(viewportWidth * parseInt(width, 10) / 100);
-  }
-  return undefined;
-}
-
-function clampColumnWidth<R, SR>(
-  width: number,
-  { minWidth, maxWidth }: Column<R, SR>,
-  minColumnWidth: number
-): number {
-  width = Math.max(width, minWidth ?? minColumnWidth);
-
-  if (typeof maxWidth === 'number') {
-    return Math.min(width, maxWidth);
-  }
-
-  return width;
+  }, [colOverscanEndIdx, colOverscanStartIdx, columns, isGroupRow, rows]);
```

Comment:
We should split this into two `useMemo`s so we can avoid recreating the `viewportColumns` array every time we change `rows`/`isGroupRow`.

`if ((colSpan + column.idx - 1) > startIdx) {` also had a bug where we would render a cell that should have in a span,

Before:
![image](https://user-images.githubusercontent.com/567105/115423788-04d49700-a1f6-11eb-971f-5ac18f4fdb62.png)
![image](https://user-images.githubusercontent.com/567105/115423821-0bfba500-a1f6-11eb-8eb8-edc0dd911173.png)
![image](https://user-images.githubusercontent.com/567105/115423968-29307380-a1f6-11eb-9bd9-c784f897c0ad.png)
^ the 7x7 cell is rendered when it shouldn't.

After:
![image](https://user-images.githubusercontent.com/567105/115423646-e9698c00-a1f5-11eb-8f50-68d66d5617e6.png)
![image](https://user-images.githubusercontent.com/567105/115423687-f38b8a80-a1f5-11eb-8a8f-a05a2c6e03e9.png)
No funky behavior even when "offscreen".

---

We also still need to loop through the header and summary spans, as it can lead to bugs right now:
![image](https://user-images.githubusercontent.com/567105/115424213-6563d400-a1f6-11eb-867d-19448e301f34.png)
![image](https://user-images.githubusercontent.com/567105/115424239-6b59b500-a1f6-11eb-8557-a58f3834988d.png)
Header for column `10` appears when the cell position for 8 and 9 are "offscreen".

---

```ts
  const startIdx = useMemo(() => {
    let startIdx = colOverscanStartIdx;
    // find the column that spans over a column within the visible columns range and adjust colOverscanStartIdx
    for (const row of rows) {
      if (isGroupRow(row)) continue;
      for (let colIdx = 0; colIdx < colOverscanStartIdx; colIdx++) {
        const column = columns[colIdx];
        const colSpan = getColSpan(column, columns, { type: 'ROW', row });
        if (!colSpan) continue;
        if ((colIdx + colSpan) > startIdx) {
          startIdx = colIdx;
        }
      }
    }

    return startIdx;
  }, [rows, colOverscanStartIdx, isGroupRow, columns]);

  return useMemo((): readonly CalculatedColumn<R, SR>[] => {
    const viewportColumns: CalculatedColumn<R, SR>[] = [];
    for (let colIdx = 0; colIdx <= colOverscanEndIdx; colIdx++) {
      const column = columns[colIdx];

      if (colIdx < startIdx && !column.frozen) continue;
      viewportColumns.push(column);
    }

    return viewportColumns;
  }, [startIdx, colOverscanEndIdx, columns]);
```

Reply:
Also we should probably just loop through the visible rows, and not all rows.